lib/sparse/src/index/search_context.rs

```diff
@@ -0,0 +1,447 @@
+            if let Some(record_id) = posting_iterator
+                .posting_list_iterator
+                .peek()
+                .map(|element| element.record_id)
+            {
+                // accumulate score for the current record id
+                if record_id == min_record_id {
+                    let element = posting_iterator.posting_list_iterator.next().unwrap();
+                    score +=
+                        element.weight * self.query.weights[posting_iterator.query_weight_offset];
+                }
+            }
+        }
+
+        Some(ScoredCandidate {
+            score,
+            vector_id: min_record_id,
+        })
+    }
+
+    pub fn next_min(to_inspect: &[IndexedPostingListIterator<'_>]) -> Option<u32> {
+        let mut min_record_id = None;
+
+        // Iterate first time to find min record id at the head of the posting lists
+        for posting_iterator in to_inspect.iter() {
+            if let Some(next_element) = posting_iterator.posting_list_iterator.peek() {
+                if Some(next_element.record_id) < min_record_id || min_record_id.is_none() {
+                    min_record_id = Some(next_element.record_id);
+                }
+            }
+        }
+
+        min_record_id
+    }
+
+    /// Make sure the longest posting list is at the head of the posting list iterators
+    pub fn sort_posting_lists_by_len(&mut self) {
+        // decreasing order
+        self.postings_iterators.sort_by(|a, b| {
+            b.posting_list_iterator
+                .len_to_end()
+                .cmp(&a.posting_list_iterator.len_to_end())
+        });
+    }
+
+    pub fn search(&mut self) -> Vec<ScoredCandidate> {
+        while let Some(candidate) = self.advance() {
+            // push candidate to result queue
+            self.result_queue.push(candidate);
+
+            // we potentially have enough results to prune low performing posting lists
+            if self.result_queue.len() == self.top {
+                // current min score
+                let min_score = self.result_queue.top().unwrap().score;
+
+                // sort posting lists by length to try to prune the longest one
+                self.sort_posting_lists_by_len();
+
+                self.prune_longest_posting_list(min_score);
+            }
+        }
+        // posting iterators exhausted, return result queue
+        let queue = std::mem::take(&mut self.result_queue);
+        queue.into_vec()
+    }
+
+    /// Prune posting lists that cannot possibly contribute to the top results
+    /// Assumes longest posting list is at the head of the posting list iterators
+    /// Returns true if the longest posting list was pruned
+    pub fn prune_longest_posting_list(&mut self, min_score: f32) -> bool {
+        // compute skip target before acquiring mutable reference to posting list iterator
+        let skip_to = if self.postings_iterators.len() == 1 {
```

Comment:
Would it make sense to check `<= 1` here?

Reply:
This is good observation, the search routine must protect itself from sparse query vectors not mathcing any known posting list in the index.

Added this https://github.com/qdrant/qdrant/pull/2318/commits/c862c4967afd63b46e0b977432b7565f793121db

Reply:
Fantastic!

Reply:
To prevent direct access to `prune_longest_posting_list` bypassing the check, we might want to make this method (and possibly another) private:

```patch
diff --git a/lib/sparse/src/index/search_context.rs b/lib/sparse/src/index/search_context.rs
index 6e980ed6..58221c9e 100644
--- a/lib/sparse/src/index/search_context.rs
+++ b/lib/sparse/src/index/search_context.rs
@@ -98,7 +98,7 @@ impl<'a> SearchContext<'a> {
         })
     }
 
-    pub fn next_min(to_inspect: &[IndexedPostingListIterator<'_>]) -> Option<u32> {
+    fn next_min(to_inspect: &[IndexedPostingListIterator<'_>]) -> Option<u32> {
         let mut min_record_id = None;
 
         // Iterate first time to find min record id at the head of the posting lists
@@ -114,7 +114,7 @@ impl<'a> SearchContext<'a> {
     }
 
     /// Make sure the longest posting list is at the head of the posting list iterators
-    pub fn sort_posting_lists_by_len(&mut self) {
+    fn sort_posting_lists_by_len(&mut self) {
         // decreasing order
         self.postings_iterators.sort_by(|a, b| {
             b.posting_list_iterator
```